### DISCUSSION TITLE
parallel foreach on zip entries

### DISCUSSION THREAD

**Discussion by eilat-inbal** [icsharpcode/SharpZipLib#718] [Q&A]
Hi,
I am trying to extract the zip entries in a parallel foreach, and encountered a weird behavior. Can anyone try and explain it?
when the zip is using a ZipCrypto encryption- this works really well and the performance is improved (4 times faster!).
but when the zip is using AES256- the performance is not improved, and is even slower than a regular, non-parallel loop sometimes.

here is my loop:
```cs
Parallel.ForEach(zf.OfType<ICSharpCode.SharpZipLib.Zip.ZipEntry>(), zipEntry =>
{
    byte[] buffer = new byte[4096];
    Stream zipStream = zf.GetInputStream(zipEntry);
    string fullZipToPath = Path.Combine(destFolder, zipEntry.Name);
    using (FileStream streamWriter = File.Create(fullZipToPath))
    {
        StreamUtils.Copy(zipStream, streamWriter, buffer);
    }
});
```

I  hope there is a way to parallelize the zip entries extraction in AES256!
Or if there are other ways to improve the performance, I would love to hear!
thanks,
Inbal

**Comment by piksel** [comment]
You could try using multiple memory streams containing the entire zipfile and see if there is a difference. It might be I/O-related, or it might be due to your runtimes AES implementation not being parallelizable. 🤷‍♀️
SharpZipLib uses [`Aes.Create()`](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.aes.create?view=net-6.0#system-security-cryptography-aes-create) to get the default implementation.
